### PR TITLE
RELEASE.md rewrite

### PR DESCRIPTION
**Pull request by Lorak-mmk**
RELEASE.md contained outdated procedures which were not useful to a person that wanted to perform a release.
To avoid explaining manually how to do a release again and again I've rewritten this file with current information about release process
and other maintenance-related stuff with much more detail than before.

It should aid current and future maintainers during any releases.

I've also changed triggers on docs generation workflow - commit message explains why.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] ~~I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~~I have provided docstrings for the public items that I want to introduce.~~
- [ ] ~~I have adjusted the documentation in `./docs/source/`.~~
- [ ] ~~I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by Lorak-mmk** [comment]
@piodul @wprzytula Please read the doc and verify that all information are correct.

@muzarski @dkropachev as you are not mainteiners I'd appreciate if you read the doc to check that everything in understandable for person not familiar with maintenance processes of this project.

File: RELEASE.md
Comment (muzarski) [inline RELEASE.md]: `or some branch-* branch that updates something in `docs/` directory` - is this part true after the previous commit?

**Review by muzarski — COMMENTED** [review]
Really helpful PR. I'm glad you provided the examples of some previous releases (with corresponding links to PRs/commits).

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: `For a source` or `For sources`

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: Subtlety: `can't` -> `mustn't`, to stress that it's the releaser's responsibility to ensure that.

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: What are those 3 main crates? I don't think you defined them.

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: nit: `crates.io` (trailing dot)

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: By `+` in `TAGS + BRANCHES` you mean set union?

Comment (wprzytula) [reply RELEASE.md]: If so, then just for clarity let's write `\union` instead, or use the dedicated character: `∪`

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: simiar -> similar

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: Let's make this bold to stand out.

Comment (wprzytula) [reply RELEASE.md]:
```md
It however means that **any release requires updating `conf.py` on `main` branch** to have the new version rendered on documentation book.
```

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: I think `Release 0.13.2` is missing after the fix D is backported, because at the same time `Release 0.14.1` is done with fix D and this way we can misleadingly suggest that some fixes on some branches may not be released (potentially never).

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: Real life examples would be appreciated here. I suggest:
- one example for cherry-picking a whole PR,
- one for a single-commit PR,
- one for a single commit from a multi-commit PR.

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: nit: `a write access` -> `write access` (abstract noun)

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: `chery-pick` -> `cherry-pick`

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: `Releasing a minor version`

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: Let's make this warning stand out:

Comment (wprzytula) [reply RELEASE.md]: `**NOTE: Don't merge this PR using GitHub interface - just wait for approval.**`

Comment (wprzytula) [reply RELEASE.md]: ...and optionally provide a short explanation why merging using the GH UI is wrong.

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: `(optional, but recommended)` missing here

File: RELEASE.md
Comment (wprzytula) [inline RELEASE.md]: let's replace exact numbers with some placeholders not to forget updating them when using the template.

File: RELEASE.md
Comment (wprzytula) [reply RELEASE.md:9]: Ah, it's removed in the next commit.

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md:9]: Having some automation would be awesome and we should do that - but it is out of scope of this PR. In this PR I wanted to document the current process.

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md]: Those are Python lists - so I meant a concatenation of them. It should be identical to set union, unless someone decides to have a git branch and tag with a same name, which I hope we never do.

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md]: Thanks for noticing, I apprently pasted a wrong version of the graph.

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md]: I'll add this.

I'm just unable to understand how it is possible that GitHub doesn't support fast-forward merges in UI. It's absurd.

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md]: I left the numbers as-is so that the format of them is easier to understand,

Comment (wprzytula) [reply RELEASE.md]: So maybe let's add a TODO comment closely?

File: RELEASE.md
Comment (Lorak-mmk) [reply RELEASE.md:9]: > Ah, it's removed in the next commit.

I'll move the commit that removes the script before the commit that rewrites RELEASE.md

**Comment by Lorak-mmk** [comment]
v2:
- Addressed review comments
- Added section about writing release notes: what to include and how, how to count commits by author etc

File: MAINTENANCE.md
Comment (wprzytula) [inline MAINTENANCE.md:231]: Embrace the weirdness!

File: MAINTENANCE.md
Comment (wprzytula) [inline MAINTENANCE.md:217]: Let's consider the following scenario. A user decides to update the driver to a next major version (e.g. 0.14). Thus, they read the release notes for that release. However, as minor dependency updates happens automatically with Cargo, they miss the release notes for the previous minor release (0.13.2) that brought some crucial changes/bugfixes.

To prevent such scenario, let's include all changes that happened since last major release in next major release's notes.

**Review by muzarski — APPROVED** [review]

File: MAINTENANCE.md
Comment (Lorak-mmk) [reply MAINTENANCE.md:217]: It will cause duplication of entries, but will make it easier to write release notes. I think this tradeoff is acceptable.

Comment (Lorak-mmk) [reply MAINTENANCE.md:217]: Of course we will need to change a lot of those procedures / rules when 1.0 is released.

File: MAINTENANCE.md
Comment (Lorak-mmk) [reply MAINTENANCE.md:231]: No more weirdness after I changed the procedure according to your comment below - so this note will be removed

**Comment by Lorak-mmk** [comment]
v3: Addressed review comments, changed the release notes guidelines according to @wprzytula suggestion

**Review by wprzytula — APPROVED** [review]
Great job! Now every release is going to be easier, with everything neatly written down and explained.

**Pull request merged**
